### PR TITLE
Fixes #87. Ensure deactivate.sh works if activate.sh is not run

### LINKED ISSUE #87: openjdk cannot be installed through bash script using `-u`
Edit: the issue below was caused by the `-u` flag in bash, see the post below. So my personal immediate problem is solved by simply not using that flag. However, it might be worthwhile to fix the problem in the openjdk feedstock anyway.

---

- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue:

It seems not possible to install openjdk through a bash script.

```
$ cat openjdktest.sh
set -Eeuo pipefail
set -x
# one of the below is necessary:
# eval "$(conda shell.bash hook)"
source "$CONDA_PREFIX/etc/profile.d/conda.sh"
conda create -y -n openjdktest
conda activate openjdktest
conda install -c conda-forge -y openjdk

$ bash openjdktest.sh 
...
  zstd               conda-forge/linux-64::zstd-1.5.0-ha95c52a_0


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
+ __conda_reactivate
+ local ask_conda
++ PS1='(openjdktest) '
++ __conda_exe shell.posix reactivate
++ __add_sys_prefix_to_path
++ '[' -n '' ']'
+++ dirname /home/.../bin/conda
++ SYSP=/home/.../bin
+++ dirname /home/.../bin
++ SYSP=/home/...
++ '[' -n '' ']'
++ PATH=/home/.../bin:/home/.../envs/openjdktest/bin:/home/.../condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
++ export PATH
++ /home/.../bin/conda shell.posix reactivate
+ ask_conda='. "/home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh"
PS1='\''(openjdktest) '\''
export PATH='\''/home/.../envs/openjdktest/bin:/home/.../condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin'\''
export CONDA_SHLVL='\''2'\''
export CONDA_PROMPT_MODIFIER='\''(openjdktest) '\''
. "/home/.../envs/openjdktest/etc/conda/activate.d/openjdk_activate.sh"'
+ eval '. "/home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh"
PS1='\''(openjdktest) '\''
export PATH='\''/home/.../envs/openjdktest/bin:/home/.../condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin'\''
export CONDA_SHLVL='\''2'\''
export CONDA_PROMPT_MODIFIER='\''(openjdktest) '\''
. "/home/.../envs/openjdktest/etc/conda/activate.d/openjdk_activate.sh"'
++ . /home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh
/home/.../envs/openjdktest/etc/conda/deactivate.d/openjdk_deactivate.sh: line 1: JAVA_HOME_CONDA_BACKUP: unbound variable
```

That is, it seems that `conda install` calls `__conda_reactivate`, which calls `openjdk_deactivate.sh`, which fails because for some reason `openjdk_activate.sh` is never ran.

Note that performing the commands in ` openjdktest.sh` directly in the shell itself will work. See also https://stackoverflow.com/questions/52779016/conda-command-working-in-command-prompt-but-not-in-bash-script/52813960

Arguably this is not directly a problem caused by openjdk feedstock itself, but rather an inconsistency within conda. ~However, at this moment I do not know how to install openjdk through a bash script. As in, I do not yet have a workaround for this problem, so I'm posting this issue here.~ And it seems that the problem can be solved by having the deactivate script only reset the environment variables if the backup variables actually have been set.

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /home/...:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
beautifulsoup4            4.10.0             pyha770c72_0    conda-forge
brotlipy                  0.7.0           py38h497a2fe_1003    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2021.10.8            ha878542_0    conda-forge
certifi                   2021.10.8        py38h578d9bd_1    conda-forge
cffi                      1.14.6           py38h400218f_0  
chardet                   4.0.0            py38h578d9bd_2    conda-forge
charset-normalizer        2.0.0              pyhd8ed1ab_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
conda                     4.10.3           py38h578d9bd_4    conda-forge
conda-build               3.21.6           py38h578d9bd_0    conda-forge
conda-package-handling    1.7.3            py38h497a2fe_1    conda-forge
cryptography              35.0.0           py38h3e25421_2    conda-forge
filelock                  3.4.0              pyhd8ed1ab_0    conda-forge
glob2                     0.7                        py_0    conda-forge
gprof2dot                 2019.11.30                 py_0    conda-forge
icu                       69.1                 h9c3ff4c_0    conda-forge
idna                      3.1                pyhd3deb0d_0    conda-forge
jinja2                    3.0.3              pyhd8ed1ab_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libarchive                3.5.2                hccf745f_1    conda-forge
libffi                    3.3                  h58526e2_2    conda-forge
libgcc-ng                 11.2.0              h1d223b6_11    conda-forge
libgomp                   11.2.0              h1d223b6_11    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
liblief                   0.11.5               h9c3ff4c_1    conda-forge
libstdcxx-ng              11.2.0              he4da1e4_11    conda-forge
libxml2                   2.9.12               h885dcf4_1    conda-forge
libzlib                   1.2.11            h36c2ea0_1013    conda-forge
lxml                      4.6.2                    pypi_0    pypi
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
lzo                       2.10              h516909a_1000    conda-forge
markupsafe                2.0.1            py38h497a2fe_1    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
patchelf                  0.13                 h58526e2_0    conda-forge
pkginfo                   1.8.1              pyhd8ed1ab_0    conda-forge
psutil                    5.8.0            py38h497a2fe_2    conda-forge
py-lief                   0.11.5           py38h709712a_1    conda-forge
pycosat                   0.6.3           py38h497a2fe_1009    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1            py38h578d9bd_4    conda-forge
python                    3.8.5                h7579374_1  
python-libarchive-c       3.1              py38h578d9bd_1    conda-forge
python_abi                3.8                      2_cp38    conda-forge
pytz                      2021.3             pyhd8ed1ab_0    conda-forge
pyyaml                    6.0              py38h497a2fe_3    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
requests                  2.26.0             pyhd8ed1ab_0    conda-forge
ripgrep                   13.0.0               habb4d0f_1    conda-forge
ruamel_yaml               0.15.80         py38h497a2fe_1006    conda-forge
setuptools                59.2.0           py38h578d9bd_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
soupsieve                 2.3                pyhd8ed1ab_0    conda-forge
sqlite                    3.36.0               h9cd32fc_2    conda-forge
svg2tikz                  1.0.0.dev0               pypi_0    pypi
tk                        8.6.11               h27826a3_1    conda-forge
tqdm                      4.62.3             pyhd8ed1ab_0    conda-forge
urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
yaml                      0.2.5                h516909a_0    conda-forge
zlib                      1.2.11            h36c2ea0_1013    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge

```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
     active environment : base
    active env location : /home/.../miniconda3
            shell level : 1
       user config file : /home/.../.condarc
 populated config files : /home/.../.condarc
          conda version : 4.10.3
    conda-build version : 3.21.6
         python version : 3.8.5.final.0
       virtual packages : __cuda=11.2=0
                          __linux=5.4.0=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/.../miniconda3  (writable)
      conda av data dir : /home/.../miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/.../miniconda3/pkgs
                          /home/.../.conda/pkgs
       envs directories : /home/.../miniconda3/envs
                          /home/.../.conda/envs
               platform : linux-64
             user-agent : conda/4.10.3 requests/2.26.0 CPython/3.8.5 Linux/5.4.0-90-generic ubuntu/20.04.3 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</details>

### PR DESCRIPTION
**Pull request by hugobuddel**
It can happen that the deactivate script is run when the activate script has not been run. The deactivate script should therefore be robust in verifying all relevant variables are set.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Closes #87

### CODE REVIEW

**Comment by hugobuddel** [comment]
I'm building the package now, and will try to replicate the two problems described in #139 once the package is built.

Only the bash scripts have been changed. Perhaps the two sets of windows scripts should be updated as well? I have no convenient way of testing those though.

**Comment by hugobuddel** [comment]
This indeed seems to work. E.g., combining the two tests in #87 :
- performing the installation in a bash script with `set -u`
- verifying `$JAVA_HOME`  after deactivation (after installation, and after activation

`openjdktest3.sh` :

```
set -Eeuo pipefail

JAVA_HOME=/hello/there
echo "At the start JAVA_HOME is $JAVA_HOME"

source "$CONDA_PREFIX/etc/profile.d/conda.sh"
conda create -y -n openjdktest
conda activate openjdktest
conda install -c local -y openjdk
conda deactivate

echo "Halfway through JAVA_HOME is $JAVA_HOME"

conda activate openjdktest
conda deactivate

echo "At the end JAVA_HOME is $JAVA_HOME"
```

Then running it, `bash openjdktest3.sh`:
```
At the start JAVA_HOME is /hello/there                                
Collecting package metadata (current_repodata.json): done                                                                                   
[...]
Executing transaction: done
Halfway through JAVA_HOME is /hello/there
At the end JAVA_HOME is /hello/there
```

Two conclusions:
- The script ends without errors, so both the activate and deactivate scripts work with `set -u`.
- The deactivate script successfully restores `$JAVA_HOME` in both scenarios.

So what remains is porting this to the two sets of windows scripts. That is something I will not do myself, so I consider this good enough for now.

**Comment by hugobuddel** [comment]
Oops, forgot to commit the activate script. Test still valid (or actually, now valid), because it was ran locally.

Increased the build number from 0 to 1.

**Comment by hugobuddel** [comment]
@conda-forge-admin, please rerender

File: recipe/scripts/deactivate.sh [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
+# Then restore the backup, and unset the backup.
+
+# Note that the check whether the backup is set, is essential.
+# There are situations where conda executes the deactivate
+# script without having called the activate script. Without
+# this check, the deactivate script would unset JAVA_HOME in
+# those situations.
+
+# One such situation occurs when deactivating the environment
+# after installing openjdk for the first time.
```
Comment (h-vetinari) [inline recipe/scripts/deactivate.sh:12]: Thanks for that, I was struggling to figure out legitimate scenarios where deactivation is run without prior activation.

**Review by h-vetinari — APPROVED** [review]
LGTM, thanks for the PR!

**Pull request merged**